Settings page keyboard navigation

Samq64 opened the pull request:
Resolves #2532

### Changes

- The settings page and popup are now usable with the keyboard
- Converts the more settings modal into an HTML `<dialog>` element
- Consolidates the switch styles into its own file
- Consolidates `addon-group`, `btn-dropdown` and `back-button` into a single class
- Adds a dropdown compnent with arrow key navigation
- ~~Always shows the dismiss admin message link (regular comments are unaffected)~~

### Reason for changes

It's good for accessibility and not being able to toggle addons with the keyboard is annoying.

### Tests

Tested on Chromium

Todo:
- [x] Dropdows
- [x] Scratch messaging
- [x] Cloud games (already good)
- [x] Fix enabling addons with confirmations
- [ ] Better focus borders
- [x] Re-order table rows with the keyboard? #8934
- [ ] Navigate between addons with the arrow keys?

Samq64 committed:
Use checkbox elements for addon toggles, share styles

Samq64 committed:
Fix keyboard navigation in headers and sidebar

Samq64 committed:
More settings: use native dialog element

Samq64 committed:
Convert more divs to buttons

Samq64 committed:
Format code

Samq64 commented on images/icons/x.svg:
This SVG was also rotated 360 degrees so I removed that.

mxmou replied:
The transform was probably meant to fix a rendering bug. You can remove it if it doesn't cause a visible difference in the latest versions of browsers that SA supports.

DNin01 reviewed (commented):
Did some testing with Windows Narrator—there's still room for improvement with screen reader navigation, but, with a keyboard, things are at least usable now.

Some screen reader problems:
- Many icon buttons don't have a name
- Addon setting elements aren't linked to their labels
- Item lists like the category panel don't function like lists

Some keyboard navigation problems:
- Setting tooltips can't be accessed

DNin01 committed:
Make list items tabbable

DNin01 commented:
I added keyboard navigation to dropdowns by simply setting `tabindex="0"` on them.

mxmou commented:
The correct keyboard behavior for dropdowns is that up and down arrows focus the previous/next item and Tab focuses the next element outside the dropdown.

DNin01 commented:
> The correct keyboard behavior for dropdowns is that up and down arrows focus the previous/next item and Tab focuses the next element outside the dropdown.

I have before made a library with this exact functionality, so I'll look into this.

Samq64 commented:
I have not done any screen reader testing myself yet.

> I have before made a library with this exact functionality, so I'll look into this.

I was hoping not to add another libary but it might be worth it if it needs more than a few lines of JS.

Samq64 committed:
More accessibile arrow buttons, consolidate more styles

Samq64 committed:
Temporary dropdown solution

Enter doesn't trigger @click on li elements, also close dropdown on
escape.

Samq64 committed:
Don't focus disabled addon elements

Samq64 committed:
Format code

Samq64 committed:
Dropdown component

Samq64 committed:
Dropdown arrow navigation

Thanks @Dnin01 for the code snippet!

Samq64 committed:
Format code

Samq64 committed:
Fix icon margin in more settings

DNin01 commented:
It works! The dropdowns, menus, and menu items are announced as such:

> - "Button, collapsed, has popup"
>   - "Export settings, menu item, 1 of 2"
>   - "Import settings, menu item, 2 of 2"

However, icon-buttons still need readable labels—currently most of the images don't have alt text, and when there's no text, the default name is just "graphic" or "button".

> I'm not sure if the menus should stay open when tabbing away, currently they doesn't.

On GitHub, they collapse when you tab outside.

> > If you want to figure it out, feel free to use this snippet under the same license terms of our project.
> 
> Just to be clear you mean this repository's license, right?

Correct.

Samq64 committed:
Fix toggling addons with permssions/confirmation

Samq64 committed:
Give more buttons titles, focusable setting tooltips

Samq64 committed:
Fix dropdowns not closing when clicking outside

Samq64 committed:
Format code

Samq64 committed:
Start working on Scratch messaging

Samq64 committed:
Format code

Samq64 committed:
Delete and post comments with the keyboard

Samq64 committed:
Format code

Samq64 committed:
Always show dismiss button

Samq64 committed:
Shorter dropdown title

Samq64 commented:
> However, icon-buttons still need readable labels—currently most of the images don't have alt text, and when there's no text, the default name is just "graphic" or "button".

I've added titles to more buttons but I think full screen reader support is out of the scope of this PR.

mxmou commented on _locales/en/messages.json:
```diff
@@ -176,6 +194,15 @@
     "message": "Preview"
   },
+  "addRow": {
+    "message": "Add table row"
```
I'm not sure if we should use the word table in the UI for table settings - they don't actually look like tables, so I doubt users think of them as such. I suggest using "item" instead of "table row".

mxmou commented on _locales/en/messages.json:
```diff
@@ -188,6 +215,9 @@
     "message": "Dark"
   },
+  "toggleTheme": {
+    "message": "Toggle dark mode"
```
It would be nice (and easier to translate) if there were separate "Switch to dark theme" and "Switch to light theme" messages. ("Theme", not "mode", for consistency with the text in More Settings.)

mxmou reviewed (commented):
Browsers render `<button>` text in a different font by default, which causes inconsistencies like this:

<img width="281" height="121" alt="Image" src="https://github.com/user-attachments/assets/e7053950-040f-47eb-b803-7b9b56cdd720" />

A simple fix would be
```css
button {
  font: inherit;
}
```

Samq64 committed:
Apply Maximouse's suggestions

Co-authored-by: Maximouse <brestl1c@outlook.com>

Samq64 committed:
Format code

mxmou commented on _locales/en/messages.json:
```diff
@@ -195,10 +195,10 @@
+    "message": "Add preset row"
   },
   "deleteRow": {
     "message": "Delete table row"
```
```suggestion
    "message": "Delete row"
```

Samq64 committed:
Forgot one

Co-authored-by: Maximouse <brestl1c@outlook.com>

Samq64 committed:
Minor cleanup

mxmou commented on webpages/settings/index.html:
```diff
@@ -299,7 +306,7 @@ <h1 v-cloak>{{ msg("settings") }}</h1>
       </div>
     </div>
-    <modal class="more-settings" :is-open.sync="moreSettingsOpen" :title="msg('moreSettings')" v-cloak>
+    <modal class="more-settings" :is-open.sync="moreSettingsOpen" :title="msg('moreSettings')" v-el:moreSettings>
```
The `v-cloak` attribute should stay - otherwise the content of the modal can briefly appear at the bottom of the popup after switching from a different tab to Addons:

<img width="442" height="645" alt="Image" src="https://github.com/user-attachments/assets/1dee962a-d508-4de1-b001-3b2ab999f7d3" />

mxmou commented on webpages/settings/style.css:
```diff
@@ -133,22 +137,28 @@ h1 {
 }
 
-.back-button:hover {
+.arrow-button:hover {
```
```suggestion
.arrow-button:hover,
.clickable-area:hover .arrow-button,
.addon-group:hover .arrow-button {
```

Samq64 committed:
Apply suggestions from code review

Co-authored-by: Maximouse <brestl1c@outlook.com>

Samq64 committed:
Revert unnecessary messaging changes

Samq64 committed:
Fix bugs

Samq64 committed:
Format code

Samq64 committed:
Consistent disabled state naming

Samq64 committed:
margin-inline consistency with master

Samq64 committed:
Fix split button highlight

mxmou reviewed (commented):
There's one remaining bug: on Firefox, the focus outlines of `.filter-option`s and popup tabs aren't fully visible because the parent has `overflow: hidden`. A negative `outline-offset` would fix that.

Samq64 committed:
Fix tab and category oulines

And popup shift when focusing the bottom Scratch messaging links

Samq64 committed:
Apply to header as well

mxmou reviewed (approved):
`.filter-option` still needs a negative `outline-offset`.

Samq64 committed:
filter-option as well

DNin01 commented:
> Navigate between addons with the arrow keys?

<details><summary>I managed to get this working, but it relies on the addon cards' style attributes being an exact value (to skip over addons in collapsed groups).</summary>

<p>

`addon-body.html`
```diff
  <template>
- <div class="addon-body" v-show="shouldShow" :id="'addon-' + addon._addonId">
+   <div
+     class="addon-body"
+     v-show="shouldShow"
+     :id="'addon-' + addon._addonId"
+     tabindex="0"
+     @keydown="handleKeys"
+   >
      <div class="addon-topbar">
```

`addon-body.js`
```diff
        },
+       handleKeys(e) {
+         if (!e.target.classList.contains("addon-body")) return;
+         if (e.ctrlKey || e.metaKey || e.altKey) return;
+         const cards = document.querySelectorAll(".addons-container .addon-body:not([style='display: none;'])");
+         const index = Array.prototype.findIndex.call(cards, (i) => i === document.activeElement);
+         if (e.key === "ArrowUp") {
+           if (!cards[index - 1]) return;
+           e.preventDefault();
+           cards[index - 1].focus();
+         } else if (e.key === "ArrowDown") {
+           if (!cards[index + 1]) return;
+           e.preventDefault();
+           cards[index + 1].focus();
+         }
+       },
        toggleAddonRequest(event) {
```

</p>

There's probably a better way to do this.</details>

DNin01 commented:
The only change needed to make badge tooltips reveal with the keyboard is:

`addon-tag.html`
```diff
  <template>
    <div
      class="badge"
      v-if="shouldShow"
+     :tabindex="tagInfo.tooltipText ? 0 : -1"
      :class="{
        'tooltip': tagInfo.tooltipText,
```

And...

<img width="340" height="158" alt="Recommended label in settings page" src="https://github.com/user-attachments/assets/aada3be4-fbb0-461c-b1d4-d0ff9a0c1675" />

Samq64 committed:
Tabbable tags, autofocus first dropdown item

Samq64 committed:
Trun select options into radio input elements

Not sure if this is any better than just manully handling keypresses

DNin01 committed:
Refine radio buttons

- Fix borders
- Fix Windows Narrator cursor
- Designate parent element as a radiogroup

DNin01 commented:
> Not sure if this is any better than just manully handling keypresses

That's even better! The options are announced as radio buttons by the screen reader.

Samq64 committed:
Apply to other sidebar toggle button

Samq64 committed:
Format code

DNin01 reviewed (approved):
Everything so far looks OK to me now. The remaining todos could easily be saved for another PR, if you want.

Samq64 committed:
Focus before button when shift-tabbing in dropdown

mxmou commented on webpages/settings/components/addon-setting.html:
```diff
@@ -73,15 +73,18 @@
-        >
-          {{ option.name }}
-        </button>
+        <div v-for="option of setting.potentialValues">
```
Can this outer div be a `template` instead? That would fix the label's `flex-grow` style not being applied, causing the options to look like this when stacked vertically (in the popup):

<img width="432" height="90" alt="image" src="https://github.com/user-attachments/assets/6c886650-bca8-4371-8707-479004b758ad" />

mxmou commented on webpages/styles/components/filter.css:
```diff
@@ -44,7 +44,7 @@
   cursor: pointer;
 }
-.filter-option:last-child {
+.filter-options > :last-child > .filter-option {
```
This change can then be reverted.

Samq64 committed:
flex-grow on div

DNin01 commented:
Now that this is done, I'll try to work on improving screen reader support and some other more specific UI interactions.
